### ISSUE TITLE
Using if to narrow down int range

### ISSUE THREAD

**Issue by still-dreaming-1** [vimeo/psalm#6778]
This issue cropped up for me in the code I work on after updating to the new release. Here is a simplified version of the issue.
https://psalm.dev/r/7db09cea09

I also tried using int ranges instead, but the issue remains.
https://psalm.dev/r/9ccc816956

It's not actually a problem for me as I was able to resolve it by simplifying the code. I'm just reporting this in case it is helpful since it seemed to work before.
https://psalm.dev/r/4f1bcd3b7f

**Comment by orklah** [comment]
I tried to solve this but it's hard.

I'm not even exactly sure what is going on so I'll left some context (possibly for future - more knowledgeable - me).
Analysing `||` seems hard. Psalm somehow transform `$a === 0 || $a > 0` into three assertions at one point `!=int(0)`, `=isset`, `positive-numeric`. This is correct and it ends up in AssertionReconciler. After a fix I made and that I gonna PR, AssertionReconciler manages to reduce this to `int<1, 1>|int<1, max>` which is mostly correct. But then I lost track of what is done.
It seems at some point Psalm loose the info or replace it with less precise type

I also note PHPStan has the same exact issue: https://phpstan.org/r/1f64d5d9-7f82-4cf9-aa31-d3033a0adb49 so it may be something really deep
